lib/segment/src/index/query_optimization/condition_converter.rs

```diff
@@ -38,19 +38,40 @@ pub fn condition_converter<'a>(
                     })
                 })
             }),
-        // ToDo: It might be possible to make this condition faster by using index to check
-        //       if there is any value. But if value if not found,
-        //       it does not mean that there are no values in payload
-        Condition::IsEmpty(is_empty) => Box::new(move |point_id| {
-            payload_provider.with_payload(point_id, |payload| {
-                check_is_empty_condition(is_empty, &payload)
-            })
-        }),
-        Condition::IsNull(is_null) => Box::new(move |point_id| {
-            payload_provider.with_payload(point_id, |payload| {
-                check_is_null_condition(is_null, &payload)
-            })
-        }),
+        // We can use index for `is_empty` condition effectively only when it is not empty.
+        // If the index says it is "empty", we still need to check the payload.
+        Condition::IsEmpty(is_empty) => {
+            let fallback = Box::new(move |point_id| {
+                payload_provider.with_payload(point_id, |payload| {
+                    check_is_empty_condition(is_empty, &payload)
+                })
+            });
+            field_indexes
+                .get(&is_empty.is_empty.key)
+                .and_then(|indexes| {
+                    indexes
+                        .first()
+                        .map(|index| get_is_empty_checker(index, fallback.clone()))
+                })
+                .unwrap_or(fallback)
```

Comment:
You can avoid cloning lambda by re-structuring into:
```suggestion
            let first_field_index = field_indexes
                .get(&is_empty.is_empty.key)
                .and_then(|indexes| indexes.first());

            let fallback = Box::new(move |point_id| {
                payload_provider.with_payload(point_id, |payload| {
                    check_is_empty_condition(is_empty, &payload)
                })
            });

            match first_field_index {
                Some(index) => get_is_empty_checker(index, fallback),
                None => fallback,
            }
```

Reply:
AH! much better